include/revng/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1083 @@
+  /// that the ABI does not allow aggregate return values to use GPRs.
+  ///
+  /// \note: If a return value doesn't fit into the specified registers or
+  /// uses irregular padding, the registers are not used and the object is
+  /// passed using the memory (stack, pointer-to-copy, etc.).
+  static constexpr size_t MaximumGPRsPerAggregateReturnValue = 0;
+
+  /// States the maximum number of GPRs available to pass a single scalar
+  /// (`int`, `__int128`, pointer, etc.) argument, meaning that it can only be
+  /// passed in the GPRs if `MaximumGPRsPerScalarArgument` is less than or
+  /// equal to the number of the registers required to fit the object.
+  ///
+  /// \note: If `MaximumGPRsPerScalarArgument` is equal to 0, it means that
+  /// the ABI does not allow scalar arguments to use GPRs.
+  ///
+  /// \note: If an argument doesn't fit into the specified registers or
+  /// uses irregular padding, the registers are not used and the object is
+  /// passed using the memory (stack, pointer-to-copy, etc.).
+  static constexpr size_t MaximumGPRsPerScalarArgument = 0;
+
+  /// States the maximum number of GPRs available to return a single scalar
+  /// (`int`, `__int128`, pointer, etc.) value, meaning that it can only be
+  /// returned in the GPRs if `MaximumGPRsPerScalarReturnValue` is less than
+  /// or equal to the number of the registers required to fit the object
+  /// including padding.
+  ///
+  /// \note: If `MaximumGPRsPerScalarReturnValue` is equal to 0, it means
+  /// that the ABI does not allow scalar return values to use GPRs.
+  ///
+  /// \note: If a return value doesn't fit into the specified registers or
+  /// uses irregular padding, the registers are not used and the object is
+  /// passed using the memory (stack, pointer-to-copy, etc.).
+  static constexpr size_t MaximumGPRsPerScalarReturnValue = 0;
+
+  /// States the required size for a single stack argument in bytes.
+  static constexpr size_t MinimumStackArgumentSize = 1;
+
+  /// Stores the list of general purpose registers allowed to be used for
+  /// passing arguments and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeArgumentRegisters = {};
+
+  /// Stores the list of general purpose registers allowed to be used for
+  /// returning values and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeReturnValueRegisters = {};
+
+  /// Stores the list of vector registers allowed to be used for passing
+  /// arguments and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    VectorArgumentRegisters = {};
+
+  /// Stores the list of vector registers allowed to be used for returning
+  /// values and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    VectorReturnValueRegisters = {};
```

Comment:
You seem to consistently prefer "vector" terminology as opposed to "float".
Can you explain what's the reasoning behind this choice?

Reply:
Once upon the time, there were pure "float" registers. But those days are gone and now the co-processor-like stuff is shared between the vector instructions and float instructions.

Mostly I prefer it this way because all the vector extensions I know of support floating point stuff as well, but not the other way around.

Reply:
So all architectures that have registers dedicated to floating point values also use them for vector operations?
There are no architecture with floating point-specific registers that have no vector extensions?

Reply:
It's the other way around, all the architectures that have dedicated vector registers also use them for floating point stuff. But there are older floating-point implementations with no vector support.

Reply:
Mmmh, so why don't we have `FloatingPointArgumentRegisters`? Is it due to the fact that we postponed floating point stuff or what?

Reply:
Yes, pretty much. If I'm not forgetting anything, the only relevant "floating point only" register set are those ~~stupid~~ old `st0-st7` stack-like registers. And those would need some special treatment either way. for everything else vector registers should suffice.

Reply:
ACK.